### PR TITLE
FilesAdapter for Azure Blob Storage

### PR DESCRIPTION
**Pull request by aneeshd16**
This is a FilesAdapter for uploading, accessing and deleting files on Azure Blob Storage.
It is very similar to how the S3 adapter works.

Use the `AzureBlobStorageAdapter` as the `filesAdapter` when initializing parse-server.
```
options.filesAdapter = new AzureBlobStorageAdapter(
    	'<connection-string>',
    	'<container-name>',
    	{} 	
);
```

or with more options:
```
options.filesAdapter = new AzureBlobStorageAdapter(
    	'<storage-account-name>',
    	'<container-name>',
    	{
            storageAccessKey: <storage-access-key>,
            host: <host>,
            directAccess: <true|false>
        } 	
);
```

IMO, this addition is needed since parse-server already supports direct deployment to Azure.

This is my first PR at a major repo, I am open to feedback and suggestions! Thanks!

### CODE REVIEW

File: src/Adapters/Files/AzureBlobStorageAdapter.js [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+import { FilesAdapter } from './FilesAdapter';
+
+export class AzureBlobStorageAdapter extends FilesAdapter {
+    // Creates an Azure Storage client.
+    // Provide storage account name or storage account connection string as first parameter
+    // Provide container name as second parameter
+    // If you had provided storage account name, then also provide storage access key
+    // Host is optional, Azure will default to the default host
+    // directAccess defaults to false. If set to true, the file URL will be the actual blob URL
+    constructor(
```
Comment (flovilmart) [inline src/Adapters/Files/AzureBlobStorageAdapter.js]: would be more consistent with other adapters to only have one parameter `options`

**Comment by flovilmart** [comment]
@aneeshd16 that looks great, unit tests would be nice too, even if they fail as you don't want to provide your keys there  :)

we were also talking with @drew-gross about the future of those adapters and we both are in favour to remove them from the core repository, and reference them through dependencies.

Moving to a dependency would also help for the testing as you'll be able to put encrypted keys in travis for the unit tests to run and check for failures.

I'm working on a PR (#549) that will enable a simpler and more flexible adapter architecture.

**Comment by aneeshd16** [comment]
Thanks for the review! Yes, I'll add unit tests soon. Looking forward to the new design.

**Comment by flovilmart** [comment]
please rebase and add the according tests into https://github.com/ParsePlatform/parse-server/blob/master/spec/FilesController.spec.js

**Comment by aneeshd16** [comment]
Okay will do!

**Comment by flovilmart** [comment]
Thx!

**Comment by felixrieseberg** [comment]
Oh man, we just saw the pull request - @aneeshd16, this is awesome! We at Microsoft made github.com/felixrieseberg/parse-server-azure-storage, but it'd certainly be great to have some tests (and for this to be more integrated with the Parse-Server development).

**Comment by aneeshd16** [comment]
This is great! Tests are on the way. < 24hrs

File: src/Adapters/Files/AzureBlobStorageAdapter.js [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        });
+    }
+
+    // Generates and returns the location of a file stored in Azure Blob Storage for the given request and filename
+    // The location is the direct Azure Blob Storage link if the option is set, otherwise we serve the file through parse-server
+    getFileLocation(config, filename) {
+        if (this._directAccess) {
+            return `http://${this._storageAccountName}.blob.core.windows.net/${this._container}/${filename}`;
+        }
+        return (config.mount + '/files/' + config.applicationId + '/' + encodeURIComponent(filename));
```
Comment (aneeshd16) [inline src/Adapters/Files/AzureBlobStorageAdapter.js]: @flovilmart Tests fail because of this. `config.mount` is `undefined` here.

**Comment by aneeshd16** [comment]
@flovilmart Added the `mount` parameter here https://github.com/ParsePlatform/parse-server/blob/master/spec/FilesControllerTestFactory.js#L8
All tests pass!

**Comment by onosol** [comment]
Is this getting merged into a new version soon? With this merged, I'm ready to go with a full Azure deployment :) Just curious what the status is.....thanks

**Comment by aneeshd16** [comment]
Yep, me too. Just tell me what needs to be done so that this can be merged!

**Comment by felixrieseberg** [comment]
@onsol: In the meantime, we at Microsoft made [an adapter you can use today](https://github.com/felixrieseberg/parse-server-azure).

```
var ParseServer         = require('parse-server').ParseServer;
var AzureStorageAdapter = require('parse-server-azure').FilesAdapter;

var account = 'YOUR_AZURE_STORAGE_ACCOUNT_NAME';
var container = 'YOUR_AZURE_STORAGE_CONTAINER_NAME';
var options = {
    accessKey: 'YOUR_ACCESS_KEY',
    directAccess: false // If set to true, files will be served by Azure Blob Storage directly
}

var api = new ParseServer({
  appId: process.env.APP_ID || 'myAppId',
  serverURL: process.env.SERVER_URL || 'http://localhost:1337'
  (...)
  filesAdapter: new AzureStorageAdapter(account, container, options);
});
```

**Comment by onosol** [comment]
@felixrieseberg 
Will check it out...thanks!

**Comment by gfosco** [comment]
Like #971 I'd really like to see this become its own module on npm, instead of being included in the core server.  It's easier to test and maintain that way.  For example: https://www.npmjs.com/package/parse-server-sendgrid-adapter

**Comment by felixrieseberg** [comment]
...and @aneeshd16, we at Microsoft would love contributions over at https://www.npmjs.com/package/parse-server-azure <3!

**Comment by flovilmart** [comment]
Closing as we pimped the adapter tests and moved all adapters to external packaged

**Pull request closed without merging**